users/views.py

```diff
@@ -93,10 +96,48 @@ def get(self, request):
 
 class CustomTokenObtainPairSerializer(TokenObtainPairSerializer):
     @classmethod
-    def get_token(cls, user):
+    def get_token(cls, user) -> dict:
         token = super().get_token(user)
         token['email'] = user.email 
         return token
 
 class LoginView(TokenObtainPairView):
-    serializer_class = CustomTokenObtainPairSerializer
\ No newline at end of file
+    """
+    View for handling user login and generating JWT tokens.
+
+    This view uses the CustomTokenObtainPairSerializer to include additional
+    information in the token payload.
+    """
+    serializer_class = CustomTokenObtainPairSerializer
+
+
+
+
+
+
+class CustomJWTAuthentication(JWTAuthentication):
+    def authenticate(self, request) -> tuple:
+        """
+        Authenticate the user using JWT and attach company_id to the request.
+
+        Args:
+            request: The HTTP request object.
+
+        Returns:
+            tuple: A tuple containing the authenticated user and the token.
+
+        Raises:
+            AuthenticationFailed: If the token is invalid or missing company_id.
+        """
+        response = super().authenticate(request)
+        if response is None:
+            return None
+
+        user, token = response
+        company_id = token.get("company_id")
+
+        if not company_id:
+            raise AuthenticationFailed("Invalid token: company_id missing.")
+
+        request.company_id = company_id
```

Comment:
Directly setting `request.company_id = company_id` might override an existing request attribute, potentially leading to unexpected behavior. Use `setattr()` with a check modify:
```
request.company_id = company_id
```
To:
```
if not hasattr(request, "company_id"):
    setattr(request, "company_id", company_id)
```